### PR TITLE
Add a dirty bit to OptionsDB and only Commit() when something has cha…

### PR DESCRIPTION
**Pull request by LGM-Doyle**
…nged.

This fixes issue #618.

### CODE REVIEW

File: util/OptionsDB.h [hunk 1]
```diff
@@ -338,8 +344,23 @@ class FO_COMMON_API OptionsDB
...
     std::map<std::string, Option>   m_options;
     static OptionsDB*               s_options_db;
+    bool                            m_dirty; //< has OptionsDB changed since last Commit()
 
     friend OptionsDB& GetOptionsDB();
 };
 
+template <typename T>
+bool OptionsDB::Option::SetFromValue(const T& value_) {
+    if (value.type() != typeid(T))
```
Comment (geoffthemedio) [inline util/OptionsDB.h:354]: this makes me a bit worried about portability...

I'd probably rather try the cast within a try-catch block and react to whether bad_any_cast is thrown...

File: util/OptionsDB.h [hunk 2]
```diff
@@ -306,7 +308,11 @@ class FO_COMMON_API OptionsDB
                const boost::any& default_value_, const std::string& description_,
                const ValidatorBase *validator_, bool storable_, bool flag_, bool recognized_);
 
-        void            SetFromString(const std::string& str);
+        // SetFromValue returns true if value changed
+        template <typename T>
+        bool            SetFromValue(const T& value);
```
Comment (geoffthemedio) [inline util/OptionsDB.h]: inconsistent parameter name here...

File: util/OptionsDB.h [hunk 3]
```diff
@@ -338,8 +344,23 @@ class FO_COMMON_API OptionsDB
 
     std::map<std::string, Option>   m_options;
     static OptionsDB*               s_options_db;
+    bool                            m_dirty; //< has OptionsDB changed since last Commit()
 
     friend OptionsDB& GetOptionsDB();
 };
 
+template <typename T>
+bool OptionsDB::Option::SetFromValue(const T& value_) {
```
Comment (geoffthemedio) [inline util/OptionsDB.h:353]: ...and here

File: util/OptionsDB.h (hunk 1 above)
Comment (LGM-Doyle) [inline util/OptionsDB.h:354]: I can change this but I'd like to keep it:
1. It was moved from OptionsDB.cpp:290, so it has been in the code base for some time.
2. It is the standard idiom for boost::any, right on the front page example.  (The very next example on the same page is try{} catch(boost::bad_any_cast &) )
3. It is less verbose.

File: util/OptionsDB.h (hunk 3 above)
Comment (LGM-Doyle) [inline util/OptionsDB.h:353]: Fixed.

**Comment by LGM-Doyle** [comment]
@geoffthemedio you closed the issue, but did not merge this PR.  Was that intentional?

**Comment by ghost** [comment]
I'm getting a consistent segfault with this patch when enabling fullscreen mode.  If I am reading things correctly, validator appears to be null.

bool changed =  validator->String(value) != validator->String(value_);

Linux 64 bit, compiled from git sources.

**Comment by LGM-Doyle** [comment]
@spikethehobbitmage thanks for the bug report.  I've pushed a fix in PR #700.

**Pull request merged**